Add documentation on Semgrep findings and new Analytics page (#97)
returntocorp/semgrep-docs: docs/managing-findings.md

# Managing findings

[TOC]

> # Findings

Reviewer: I think you can make this heading an H2 (`## Findings`) and that would flatten the TOC by one level. I realize the content that follows this heading is all about findings, however so is the whole page, and that already has an H1 (“Managing findings”).

Author: Good call, I forgot about "Managing findings" above :+1:

Author: Reverted back... looks like moving to `h2` removes the `TOC`? It also looks like all other pages use another `h1` below the TOC too :man_shrugging:

Reviewer: I’m sorry, my original suggestion was incorrect. Instead of making that line Findings an H2, the correct change is to make all the section headings H1s (and the subsections become H2s):

![Screen Shot 2021-02-09 at 12 47 33 PM](https://user-images.githubusercontent.com/302941/107426372-fc186400-6ad4-11eb-97e4-1be28589e37b.png)

> A finding is the core unit of Semgrep's analysis. Findings are generated when a Semgrep rule matches a piece of code. After matching, a finding can make its way through 3 parts of the Semgrep ecosystem: [`semgrep`](https://github.com/returntocorp/semgrep), [`semgrep-action`](https://github.com/returntocorp/semgrep-action), and [semgrep.dev](https://semgrep.dev/).

Reviewer: “A finding is the core unit of Semgrep's analysis” → might it be a core result of the analysis, or core output?

Author: ```suggestion
A finding is the core result of Semgrep's analysis. Findings are generated when a Semgrep rule matches a piece of code. After matching, a finding can make its way through 3 parts of the Semgrep ecosystem: [`semgrep`](https://github.com/returntocorp/semgrep), [`semgrep-action`](https://github.com/returntocorp/semgrep-action), and [semgrep.dev](https://semgrep.dev/).
```

> A finding is the core unit of Semgrep's analysis. Findings are generated when a Semgrep rule matches a piece of code. After matching, a finding can make its way through 3 parts of the Semgrep ecosystem: [`semgrep`](https://github.com/returntocorp/semgrep), [`semgrep-action`](https://github.com/returntocorp/semgrep-action), and [semgrep.dev](https://semgrep.dev/).

Reviewer: ```suggestion
A finding is the core unit of Semgrep's analysis. Findings are generated when a Semgrep rule matches a piece of code. After matching, a finding can make its way through 3 parts of the Semgrep ecosystem: [`semgrep`](https://github.com/returntocorp/semgrep), [`Semgrep Action`](https://github.com/returntocorp/semgrep-action), and [Semgrep App](https://semgrep.dev/).
```

Author: ```suggestion
A finding is the core result of Semgrep's analysis. Findings are generated when a Semgrep rule matches a piece of code. After matching, a finding can make its way through 3 parts of the Semgrep ecosystem: [`semgrep`](https://github.com/returntocorp/semgrep), [`Semgrep Action`](https://github.com/returntocorp/semgrep-action), and [Semgrep App](https://semgrep.dev/).
```

## Semgrep CLI

> Semgrep CLI findings are a specific rule matching a piece of code. Multiple rules can match the same piece of code, even if they are effectively the same rule. For example, consider the following rule and code snippet:

Reviewer: ```suggestion
Semgrep CLI findings are produced by a specific rule matching a piece of code. Multiple rules can match the same piece of code, even if they are effectively the same rule. For example, consider the following rule and code snippet:
```

```
rules:
- id: finding-test
  pattern: $X == $X
  message: Finding test 1
  languages: [python]
  severity: WARNING
- id: finding-test
  pattern: $X == $X
  message: Finding test 2
  languages: [python]
  severity: WARNING
```

> ```

Reviewer: Maybe label this snippet as `test.py`?

Author: I see your point, but I generally find it safe to assume that "consider the following rule and code snippet" corresponds to the snippets below and the `test.yaml` and `test.py` command. Am I being too generous?

Reviewer: Probably just a me thing, I defer to you.

print(1 == 1)
```

Running Semgrep CLI produces the following findings:

```
$ semgrep --quiet --config test.yaml test.py
test.py
severity:warning rule:finding-test: Finding test 1
1:print(1 == 1)
--------------------------------------------------------------------------------
severity:warning rule:finding-test: Finding test 2
1:print(1 == 1)
```

*For more information on writing rules, see [Rule syntax](writing-rules/rule-syntax.md).*

## Semgrep Action

> Semgrep Action builds on Semgrep CLI findings to track the lifetime of an individual finding.

Reviewer: How about adding a tiny bit about Semgrep Action, like “designed to continuously scan commits and builds”?

```suggestion
Semgrep Action, designed to continuously scan commits and builds, builds on Semgrep CLI findings to track the lifetime of an individual finding.
```

Author: ```suggestion
Semgrep Action, designed to continuously scan commits and builds, improves on Semgrep CLI findings to track the lifetime of an individual finding.
```

A Semgrep Action finding is defined by a 4-tuple:

```
(rule ID, file path, syntactic context, index)
```

These pieces of state correspond to:

1. `rule ID`: the rule's ID within the Semgrep ecosystem.
1. `file path`: the filesystem path where the finding occurred.
1. `syntactic context`: the lines of code corresponding to the finding.
1. `index`: an index into identical findings within a file. This is used to disambiguate findings.

!!! note
    `syntactic context` is normalized by removing indentation, [`nosemgrep`](ignoring-findings.md#ignoring-findings-via-inline-comments) comments, and whitespace.

> These 4 pieces of information are then hashed and returned as the syntactic identifier: `syntactic_id`. This is how Semgrep Action uniquely identifies findings and tracks them across state transitions. Semgrep Action does not store or transmit code contents. The `syntactic context` is hashed using a one-way hashing function making it impossible to recover the original contents.

Reviewer: ```suggestion
These are hashed and returned as the syntactic identifier: `syntactic_id`. This is how Semgrep Action uniquely identifies findings and tracks them across state transitions. Semgrep Action does not store or transmit code contents. The `syntactic context` is hashed using a one-way hashing function making it impossible to recover the original contents.
```

Reviewer: Thank you for reinforcing that Semgrep Action does not store or transmit code contents!

## Semgrep App

Semgrep App builds on Semgrep Action findings to track state transitions and provide additional context for managing findings within your organization. Findings move between states according to their Semgrep Action `syntactic_id`, as mentioned above. A finding can occupy 3 states in Semgrep App: `OPEN`, `FIXED`, and `MUTED`.

### Finding states

Semgrep App finding states are defined as follows:

1. `OPEN`: the finding exists in the code and has not been muted.
1. `FIXED`: the finding existed in the code, and is no longer found.
1. `MUTED`: the finding has been ignored by a `nosemgrep` comment.

Findings move between states as follows:

> ![Finding state transitions](img/finding-state.png "Finding state transitions")

Reviewer: I love this diagram.

These transitions are defined as follows:

1. `Fix`: a previously identified `syntactic_id` no longer exists.
1. `Regression`: a previously fixed `syntactic_id` has been reintroduced.
1. `Mute`: a previously identified `syntactic_id` has been ignored.
1. `Unmute`: a previously muted `syntactic_id` has been unignored.

### Analytics

> Semgrep App provides analytics to measure Semgrep performance within your organization. Visit [Manage > Analytics](https://semgrep.dev/manage/findings) and use measurements like fix rate and findings over time to get the most out of your Semgrep deployment:

Reviewer: I think it would be worthwhile to make a special mention + screenshot of fix rate. Could be here or could be further down, near the screenshot that shows fixed, open, muted, blocking metrics.

Author: Fix rate in which context? The ruleset performance screenshot below includes fix rate among other things.

Reviewer: I mean to emphasize fix rate and give it some importance. Yes, it’s already in the screenshot and there are color coded cells in that column. However, I think we can give it more importance. Perhaps defining fix rate would be a good start. Currently it seems a bit buried to me (admittedly a subjective view).

Author: The "rate" for any state is simply (state / total), e.g. (fixed / (open + fixed + muted)). Should we call that out explicitly somewhere here? Or, what can we do to clear things up here?

Reviewer: Thanks, @mschwager. I think simply spelling out something like `fix rate = fixed issues / (open + fixed + muted issues)` would suffice.

Author: Great! I added a note near the bottom of the page :+1: let me know if there's a better place for it

Reviewer: Awesome, thank you!

![Blocking vs. non-blocking findings](img/sankey-diagram.png "Blocking vs. non-blocking findings")

Filter findings to drill down into specific areas:

> ![Findings filters](img/findings-filters.png "Findings filters")

Reviewer: These last three screenshots are bit hard to read. Any chance you can make their text larger? Perhaps chop off some of the whitespace on their right sides?

Author: Are you sure it's not just the image in the Github PR description? This seems reasonable to me: https://github.com/returntocorp/semgrep-docs/blob/3195425b43dafb8affbb55740aff808373f1f0fc/docs/img/ruleset-findings.png

The findings filters image is definitely lopsided, should I just update that one?

Reviewer: I think it’s not just the image in the PR description because I checked out your repo when reviewing the changes and looked at them on the built site. Here’s what I see:
![Screen Shot 2021-02-09 at 12 38 21 PM](https://user-images.githubusercontent.com/302941/107425445-d50d6280-6ad3-11eb-90c5-74ae6488d2e9.png)

And now the findings filters image is much more easy to read than it was.

Author: Tightened up the screenshots as much as I could :+1:
